docs/designs/custom_allreduce_training_loop_support.md

```diff
@@ -0,0 +1,237 @@
+            allreduce(*args, **kwargs)
+            report_batch_finished()
+        except HorovodInternalError:
+            init_horovod_if_needed()
+
+        ITER_STEP += 1
+```
+
+## Examples of Different Libraries Using ElasticDL AllReduce
+
+Users only need to make the following changes to their training loop:
+
+- Use the decorator of elastic training to wrap the code of gradient
+combination using Horovod.
+- Set objects to broadcast like model and optimizer. Because, one alive
+worker should broadcast its model and optimizer to other workers if
+AllReduce fails.
+
+### TensorFlow 1.x
+
+Using Tensorflow 1.x, we use `tf.Session` to execute the forward and backward
+computation. The training function definition likes the following
+code snippets.
+
+```python
+
+# Users should wrap their forward and backward computation
+# using ElasticDL decorator
+@elastic_allreduce
+def allreduce(session, train_op):
+    """Users should wrap the backward computation using ElasticDL
+    """
+    sess.run(train_op, feed_dict=feed_dict)
+
+
+def elastic_train(dataset):
+    dataset_iter = dataset.make_one_shot_iterator()
+    features, labels = dataset_iter.get_next()
+
+    loss = forward(features, labels)
+
+    global_step = tf.train.get_or_create_global_step()
+
+    lr = tf.Variable(base_lr * hvd.size())
+    optimizer = tf.train.GradientDescentOptimizer(lr)
+    optimizer = hvd.DistributedOptimizer(optimizer)
+    train_op = optimizer.minimize(loss, global_step=global_step)
+
+    with tf.Session(config=config) as session:
+        # ElasticDL provides ElasticBroadcastObject to set broadcast objects
+        ElasticBroadcastObject.set_session(session)
+
+        session.run(initializer)
+        step = 0
+        while True:
+            allreduce(session, train_op)
+            loss = sess.run(loss, feed_dict=feed_dict)
+            if step % 20 == 0:
+                logging.info("loss = {}".format(loss))
+            step += 1
+```
+
+### TensorFlow 2.x
+
+```python
+# Users should wrap their forward and backward computation
+# using ElasticDL decorator
+
+@elastic_allreduce
+def allreduce(tape, model, optimizer):
+    tape = hvd.DistributedGradientTape(tape)
+    grads = tape.gradient(loss, model.trainable_variables)
+    # Take care of the order of grads and vars if worker modifies
+    # `_non_embed_vars` during training.
+    optimizer.apply_gradients(
+        zip(grads, model.trainable_variables)
+    )
+
+def elastic_train(dataset):
+    """ ElasticDL will call the function to execute the training loop
+
+    Arguments:
+        dataset: tf.data.Dataset which initialized by ElasticDL
+    """
+    inputs = tf.keras.Input(shape=(28, 28), name="image")
+    outputs = Conv(10)(inputs)
+
+    model = tf.keras.Model(inputs=inputs, outputs=outputs, name="mnist_model")
+
+    optimizer = tf.optimizers.SGD(lr)
+
+    # Set object to broadcast
+    ElasticBroadcastObject.set_model(model)
+    ElasticBroadcastObject.set_optimizer(optimizer)
```

Comment:
Could you give a detailed explanation of `ElasticBroadcastObject`?

Reply:
The `ElasticBroadcastObject` is to contain objects which are needed to broadcast. I have deleted it and move it into the `ElasticAllReduceManager` in the design.